src/jit/lower.cpp

```diff
@@ -338,24 +338,96 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
     case GT_CAST:
         {
             assert(tree->gtPrev == tree->gtGetOp1());
-            NYI_IF(tree->gtOverflow(), "TYP_LONG cast with overflow");
-            switch (tree->AsCast()->CastFromType())
+            GenTree* srcOp = tree->gtGetOp1();
+            var_types srcType = tree->AsCast()->CastFromType();
+            var_types dstType = tree->AsCast()->CastToType();
+
+            if (tree->gtFlags & GTF_UNSIGNED)
             {
-            case TYP_INT:
-                if (tree->gtFlags & GTF_UNSIGNED)
+                srcType = genUnsignedType(srcType);
+            }
+
+            if (genActualTypeIsIntOrI(srcType) || varTypeIsGC(srcType))
+            {
+                if (varTypeIsUnsigned(srcType))
+                {
+                    loResult = srcOp;
+                    hiResult = comp->gtNewIconNode(0);
+
+                    comp->fgSnipNode(curStmt, tree);
+                }
+                else if (tree->gtOverflow() && varTypeIsUnsigned(dstType))
+                {
+                    //
+                    // An overflow check is needed only when casting from a signed type to ulong.
+                    // Change the cast type to uint to take advantage of the overflow check provided
+                    // by codegen and then zero extend the resulting uint to ulong.
+                    //
+
+                    loResult = tree;
+                    loResult->AsCast()->gtCastType = TYP_UINT;
+                    loResult->gtType = TYP_INT;
+
+                    hiResult = comp->gtNewIconNode(0);
+                }
+                else
                 {
+                    comp->fgInsertEmbeddedFormTemp(&tree->gtOp.gtOp1);
+
                     loResult = tree->gtGetOp1();
```

Comment:
@adiaaida `fgInsertEmbeddedFormTemp` replaces the operand, probably I should add a `srcOp = tree->gtGetOp1()` right after `fgInsertEmbeddedFormTemp` for clarity. Or just get rid of `srcOp` completely.